pkg/statemetrics/xr_state_metrics.go

```diff
@@ -0,0 +1,160 @@
+	return &XRStateRecorder{
+		client:    client,
+		log:       log,
+		frequency: frequency,
+
+		compositeExists: prometheus.NewGaugeVec(prometheus.GaugeOpts{
+			Subsystem: subSystem,
+			Name:      "composite_resource_exists",
+			Help:      "The number of composite resources that exist",
+		}, []string{"gvk", "composition"}),
+		compositeReady: prometheus.NewGaugeVec(prometheus.GaugeOpts{
+			Subsystem: subSystem,
+			Name:      "composite_resource_ready",
+			Help:      "The number of composite resources in Ready=True state",
+		}, []string{"gvk", "composition"}),
+		compositeSynced: prometheus.NewGaugeVec(prometheus.GaugeOpts{
+			Subsystem: subSystem,
+			Name:      "composite_resource_synced",
+			Help:      "The number of composite resources in Synced=True state",
+		}, []string{"gvk", "composition"}),
+		compositeComposedCount: prometheus.NewGaugeVec(prometheus.GaugeOpts{
+			Subsystem: subSystem,
+			Name:      "composite_resource_composed_count",
+			Help:      "The number of composed resources in the XR",
+		}, []string{"gvk", "composition"}),
+	}
+}
+
+// Describe sends the super-set of all possible descriptors of metrics
+// collected by this Collector to the provided channel and returns once
+// the last descriptor has been sent.
+func (r *XRStateRecorder) Describe(ch chan<- *prometheus.Desc) {
+	r.compositeExists.Describe(ch)
+	r.compositeReady.Describe(ch)
+	r.compositeSynced.Describe(ch)
+	r.compositeComposedCount.Describe(ch)
+}
+
+// Collect is called by the Prometheus registry when collecting
+// metrics. The implementation sends each collected metric via the
+// provided channel and returns once the last metric has been sent.
+func (r *XRStateRecorder) Collect(ch chan<- prometheus.Metric) {
+	r.compositeExists.Collect(ch)
+	r.compositeReady.Collect(ch)
+	r.compositeSynced.Collect(ch)
+	r.compositeComposedCount.Collect(ch)
+}
+
+// Record records the state of managed resources.
+func (r *XRStateRecorder) Record(ctx context.Context, gvk schema.GroupVersionKind) {
+	l := &unstructured.UnstructuredList{}
+	l.SetGroupVersionKind(gvk)
+	err := r.client.List(ctx, l)
+	if err != nil {
+		r.log.Info("Failed to list composite resources", "error", err)
+		return
+	}
+
+	labels := prometheus.Labels{
+		"gvk":         gvk.String(),
+		"composition": "",
+	}
+	r.compositeExists.With(labels).Set(float64(len(l.Items)))
+
+	var numReady, numSynced float64 = 0, 0
+	for _, o := range l.Items {
+		conditioned := xpv1.ConditionedStatus{}
+		if err := fieldpath.Pave(o.Object).GetValueInto("status", &conditioned); err != nil {
+			r.log.Info("Failed to get conditions of managed resource", "error", err)
+			continue
+		}
+
+		for _, condition := range conditioned.Conditions {
+			if condition.Type == xpv1.TypeReady && condition.Status == corev1.ConditionTrue {
+				numReady++
+			} else if condition.Type == xpv1.TypeSynced && condition.Status == corev1.ConditionTrue {
```

Comment:
is this really an "else"?

Reply:
`condition.Type` can only be `TypeReady` or `TypeSynced`